// expo-cli/build/commands/eas-build/build/builders/AndroidBuilder.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function _configPlugins() {
  const data = require("@expo/config-plugins");

  _configPlugins = function () {
    return data;
  };

  return data;
}

function _easBuildJob() {
  const data = require("@expo/eas-build-job");

  _easBuildJob = function () {
    return data;
  };

  return data;
}

function _path() {
  const data = _interopRequireDefault(require("path"));

  _path = function () {
    return data;
  };

  return data;
}

function _CommandError() {
  const data = _interopRequireDefault(require("../../../../CommandError"));

  _CommandError = function () {
    return data;
  };

  return data;
}

function _read() {
  const data = require("../../../../credentials/credentialsJson/read");

  _read = function () {
    return data;
  };

  return data;
}

function _AndroidCredentialsProvider() {
  const data = _interopRequireDefault(require("../../../../credentials/provider/AndroidCredentialsProvider"));

  _AndroidCredentialsProvider = function () {
    return data;
  };

  return data;
}

function _git() {
  const data = require("../../../../git");

  _git = function () {
    return data;
  };

  return data;
}

function _android() {
  const data = require("../../utils/expoUpdates/android");

  _android = function () {
    return data;
  };

  return data;
}

function _common() {
  const data = require("../../utils/expoUpdates/common");

  _common = function () {
    return data;
  };

  return data;
}

function _git2() {
  const data = require("../../utils/git");

  _git2 = function () {
    return data;
  };

  return data;
}

function _credentials() {
  const data = require("../credentials");

  _credentials = function () {
    return data;
  };

  return data;
}

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class AndroidBuilder {
  constructor(ctx) {
    this.ctx = ctx;

    _defineProperty(this, "credentials", void 0);

    _defineProperty(this, "secretEnvs", void 0);

    _defineProperty(this, "credentialsPrepared", false);
  }

  async setupAsync() {}

  async ensureCredentialsAsync() {
    this.credentialsPrepared = true;
    this.secretEnvs = await (0, _read().readSecretEnvsAsync)(this.ctx.commandCtx.projectDir);

    if (!this.shouldLoadCredentials()) {
      return;
    }

    const provider = new (_AndroidCredentialsProvider().default)(this.ctx.commandCtx.projectDir, {
      projectName: this.ctx.commandCtx.projectName,
      accountName: this.ctx.commandCtx.accountName
    }, {
      nonInteractive: this.ctx.commandCtx.nonInteractive,
      skipCredentialsCheck: this.ctx.commandCtx.skipCredentialsCheck
    });
    await provider.initAsync();
    const credentialsSource = await (0, _credentials().ensureCredentialsAsync)(provider, this.ctx.buildProfile.workflow, this.ctx.buildProfile.credentialsSource, this.ctx.commandCtx.nonInteractive);
    this.credentials = await provider.getCredentialsAsync(credentialsSource);
    return credentialsSource;
  }

  async ensureProjectConfiguredAsync() {
    const {
      projectDir,
      exp,
      nonInteractive
    } = this.ctx.commandCtx;
    const isProjectConfigured = await _configPlugins().AndroidConfig.EasBuild.isEasBuildGradleConfiguredAsync(projectDir);

    if (!isProjectConfigured) {
      throw new (_CommandError().default)('Project is not configured. Please run "expo eas:build:init" first to configure the project');
    }

    await (0, _git2().modifyAndCommitAsync)(async () => {
      if ((0, _common().isExpoUpdatesInstalled)(projectDir)) {
        await (0, _android().syncUpdatesConfigurationAsync)(projectDir, exp);
      }
    }, {
      startMessage: 'Making sure runtime version is correct on Android',
      commitMessage: 'Set runtime version in Android project',
      commitSuccessMessage: 'Successfully committed the configuration changes',
      successMessage: 'We updated the runtime version in your Android project',
      nonInteractive
    });
  }

  async configureProjectAsync() {
    const {
      projectDir,
      exp,
      nonInteractive
    } = this.ctx.commandCtx;
    await (0, _git2().modifyAndCommitAsync)(async () => {
      await _configPlugins().AndroidConfig.EasBuild.configureEasBuildAsync(projectDir);

      const easGradlePath = _configPlugins().AndroidConfig.EasBuild.getEasBuildGradlePath(projectDir);

      await (0, _git().gitAddAsync)(easGradlePath, {
        intentToAdd: true
      });

      if ((0, _common().isExpoUpdatesInstalled)(projectDir)) {
        await (0, _android().configureUpdatesAsync)(projectDir, exp);
      }
    }, {
      startMessage: 'Configuring the Android project',
      commitMessage: 'Configure Android project',
      commitSuccessMessage: 'Successfully committed the configuration changes',
      successMessage: 'We configured your Android project to build it on the Expo servers',
      nonInteractive
    });
  }

  async prepareJobAsync(archiveUrl) {
    if (!this.credentialsPrepared) {
      throw new Error('ensureCredentialsAsync should be called before prepareJobAsync');
    }

    if (this.ctx.buildProfile.workflow === _easBuildJob().Workflow.Generic) {
      return (0, _easBuildJob().sanitizeJob)(await this.prepareGenericJobAsync(archiveUrl, this.ctx.buildProfile));
    } else if (this.ctx.buildProfile.workflow === _easBuildJob().Workflow.Managed) {
      return (0, _easBuildJob().sanitizeJob)(await this.prepareManagedJobAsync(archiveUrl, this.ctx.buildProfile));
    } else {
      throw new Error("Unknown workflow. Shouldn't happen");
    }
  }

  async prepareJobCommonAsync(archiveUrl) {
    const buildCredentials = this.credentials ? {
      buildCredentials: {
        keystore: {
          dataBase64: this.credentials.keystore.keystore,
          keystorePassword: this.credentials.keystore.keystorePassword,
          keyAlias: this.credentials.keystore.keyAlias,
          keyPassword: this.credentials.keystore.keyPassword
        }
      }
    } : {};
    return {
      platform: _easBuildJob().Platform.Android,
      projectUrl: archiveUrl,
      secrets: { ...(this.secretEnvs ? {
          secretEnvs: this.secretEnvs
        } : {}),
        ...buildCredentials
      }
    };
  }

  async prepareGenericJobAsync(archiveUrl, buildProfile) {
    const projectRootDirectory = _path().default.relative(await (0, _git().gitRootDirectory)(), process.cwd()) || '.';
    return { ...(await this.prepareJobCommonAsync(archiveUrl)),
      type: _easBuildJob().Workflow.Generic,
      gradleCommand: buildProfile.gradleCommand,
      artifactPath: buildProfile.artifactPath,
      releaseChannel: buildProfile.releaseChannel,
      projectRootDirectory
    };
  }

  async prepareManagedJobAsync(archiveUrl, _buildProfile) {
    const projectRootDirectory = _path().default.relative(await (0, _git().gitRootDirectory)(), process.cwd()) || '.';
    return { ...(await this.prepareJobCommonAsync(archiveUrl)),
      type: _easBuildJob().Workflow.Managed,
      projectRootDirectory
    };
  }

  shouldLoadCredentials() {
    return this.ctx.buildProfile.workflow === _easBuildJob().Workflow.Managed || this.ctx.buildProfile.workflow === _easBuildJob().Workflow.Generic && !this.ctx.buildProfile.withoutCredentials;
  }

}

var _default = AndroidBuilder;
exports.default = _default;
